design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,228 @@
+desired settings. The BIOS Registry can be used as a source of documentation
+to modify the settings.
+
+## Design Details
+
+This proposes a new Custom Resource Definition (CRD) to store the settings
+read from Ironic along with the modified settings to write to Ironic. Since
+each vendor provides 100 or more BIOS Settings, its not practical to store
+these in the BareMetalHost CRD. The new CRD will be named
+`HostFirmwareSettings` and will consist of the following:
+
+- `currentFirmwareSettings` - the current BIOS Settings retrieved from Ironic
+  via the bios API will be stored in the Status section. This dataset includes
+  BIOS Registry information which provides a schema for the settings.
+- `writeableFirmwareSettings` - BIOS Settings to write to the host via Ironic
+  will be stored in the Spec section. If it is empty, it will be initially
+  populated with the name/value pairs from `currentFirmwareSettings` that do
+  not have the `read-only` or `unique` flag set.
+
+The BIOS Settings are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`, or when the settings
+are updated. Recent changes to Ironic (see the Ironic proposal
+[BIOS Registry](https://review.opendev.org/c/openstack/ironic-specs/+/774681)
+added an additional query of the BIOS Registry.
+
+The baremetal-operator can get the BIOS Settings from Ironic after the node
+has been moved to manageable via the `v1/nodes/{node_ident}/bios?detail=True`
+API.  This API provides a list of BIOS Settings (aka `Attributes`) in JSON and
+each setting includes the following data:
+
+- Name
+- Value
+- Registry information
+  - Type (e.g. enum, string, integer, boolean, or password)
+  - ReadOnly (a boolean indicating whether the attribute is read only)
+  - Unique (indicating the value is specific to the host where it was
+    retrieved)
+  - Allowable values for enum type attributes
+  - Minimum and maximum values (for numeric attributes)
+  - Minimum and maximum character lengths (for string attributes)
+
+Regarding the baremetal-operator states for when the data will be updated:
+
+- The node first transitions to manageable during the `Registration` state, so
+  at the end of that state `currentFirmwareSettings` will be populated.
+- Settings can be updated during the `Preparing` state, so at the end of that
+  state the settings will also be retrieved and used to update
+  `currentFirmwareSettings`.
+
+Upon retrieving the BIOS Settigs from Ironic, the baremetal-operator will
+copy the name/value pairs to the `writeableFirmwareSettings` field in
+`HostFirwareSettingsSpec` if that field is empty. Settings that have the
+`read-only` flag set will not be copied as these settings cannot be written
+to the BMC. Likewise, settings that have the `unique` flag set will also not
+be copied, these are settings like serial numbers or product ids for example,
+that should not be written to other servers.
```

Comment:
Shouldn't we also copy missing values? e.g. if there are r/w settings A. B, and C available, but only a value for B specified, I think we should copy values for A and C into the Spec. It follows that we need to do this continuously, whenever they get out of sync.

Reply:
Yes good point.  I will note that settings that aren't in the spec but are in the status will be copied to the spec. I think this only needs to be done when the status is updated, e,g. after moving the node to manageable and getting the data from Ironic.